ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/attr/Attr.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2022 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.function.udaf.attr;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.confluent.ksql.function.udaf.TableUdaf;
+import io.confluent.ksql.function.udaf.UdafDescription;
+import io.confluent.ksql.function.udaf.UdafFactory;
+import io.confluent.ksql.schema.ksql.SchemaConverters;
+import io.confluent.ksql.schema.ksql.SqlArgument;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.schema.ksql.types.SqlTypes;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Struct;
+
+@UdafDescription(
+    name = "ATTR",
+    description = "The ATTR() aggregation indicates there are multiple values, but only one "
+        + "was expected. For example, if aggregating against a table that semantically should "
+        + "have only one value for a column given a key, this aggregation enables users to "
+        + "indicate that they expect only a single value.\n\n"
+        + "If the aggregation encounters more than a single value for the expected singular "
+        + "column, the entire aggregation will return null."
+)
+public final class Attr {
+
+  private Attr() {
+    // checkstyle complains otherwise
+  }
+
+  @UdafFactory(description = "Collect as a singleton")
+  public static <T> TableUdaf<T, List<Struct>, T> createAttr() {
```

Comment:
As a note, Histogram needs to maintain the same internal structure and it uses a Map<T, Long>.  

Questions:

1.  Is one of these approaches "better" for some reason?
2.  If one is better, is it useful to standardize on them?

(I ask with https://github.com/confluentinc/ksql/issues/9140 in view as well since it'll use the same internal structure.  If one were really, really looking for refactoring options, the three functions may be able to share most of the code...)

Reply:
that makes total sense! thanks for explaining :)